### PR TITLE
Adding extension APIs for managing Azure Resources (subscriptions, resource groups, locations and SQL server)

### PR DESCRIPTION
**Pull request by llali**
Adding extension APIs for managing Azure Resources (subscriptions, resource groups, locations and SQL server). The APIs only include functions to get the list of subscriptions, resource groups, locations and a function to create a SQL server in Azure. These APIs are used in SQL projects (https://github.com/microsoft/azuredatastudio/pull/18833)

### CODE REVIEW

File: src/azure/azureController.ts [hunk 1]
```diff
@@ -279,4 +318,15 @@ export class AzureController {
 		}
 	}
 
+	public set SubscriptionClient(v: SubscriptionClient) {
```
Comment (Charles-Gagnon) [inline src/azure/azureController.ts:321]: What is this for? We don't want to use the same client for the entire azureController - many different things can call into that

File: src/azure/azureController.ts [hunk 2]
```diff
@@ -239,6 +249,35 @@ export class AzureController {
...
+						token: token
+					};
+				});
+				subscriptions = subscriptions.concat(array);
+			}
+
+			return subscriptions.sort((a, b) => (a.subscription.displayName || '').localeCompare(b.subscription.displayName || ''));
+		} catch (error) {
+
+			return [];
```
Comment (Charles-Gagnon) [inline src/azure/azureController.ts:277]: Why are you returning an empty array here? Seems reasonable to let the exception bubble up so the caller knows something went wrong.

File: src/azure/azureController.ts [hunk 3]
```diff
@@ -239,6 +249,35 @@ export class AzureController {
 		}
 	}
 
+	/**
+	 * Returns Azure subscriptions with tenant and token for each given account
+	 */
+	public async getAccountSessions(account: IAccount): Promise<mssql.IAzureAccountSession[]> {
```
Comment (Charles-Gagnon) [inline src/azure/azureController.ts:255]: Why is this here? You already have a subscription client in the azureResourceController so it seems better to just use that one

File: src/azure/azureResourceController.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import * as mssql from 'vscode-mssql';
+import * as azureUtils from './utils';
+import { TokenCredentialWrapper } from "./TokenCredentialWrapper";
+
+export class AzureResourceController {
+	private _subscriptionClient: SubscriptionClient | undefined;
+	private _resourceManagementClient: ResourceManagementClient | undefined;
+	private _sqlManagementClient: SqlManagementClient | undefined;
+
+	public set SubscriptionClient(v: SubscriptionClient) {
```
Comment (Charles-Gagnon) [inline src/azure/azureResourceController.ts:18]: Same here - we shouldn't ever be setting the client on the entire controller like this.

File: src/azure/azureResourceController.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	public set ResourceManagementClient(v: ResourceManagementClient) {
+		this._resourceManagementClient = v;
+	}
+
+	public set SqlManagementClient(v: SqlManagementClient) {
+		this._sqlManagementClient = v;
+	}
+
+	/**
+	 * Returns Azure locations for given subscription
```
Comment (Charles-Gagnon) [inline src/azure/azureResourceController.ts:31]: `for the given session`?

What is a session anyways in this context?

File: src/azure/azureResourceController.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		}
+		return undefined;
+	}
+
+	/**
+	 * Returns Azure resource groups for given subscription
+	 */
+	public async getResourceGroups(session: mssql.IAzureAccountSession): Promise<ResourceGroup[]> {
+		if (session?.subscription?.subscriptionId) {
```
Comment (Charles-Gagnon) [inline src/azure/azureResourceController.ts:63]: 1. session isn't marked as optional in the signature, which it should be if you actually intend on that being a valid case
2. Is it even a valid case for any of these to be undefined? It seems like you should be throwing an error instead of returning an empty array in that case so the user knows something was wrong

File: src/services/azureAccountService.ts [hunk 7]
```diff
@@ -21,10 +21,19 @@ export class AzureAccountService implements mssql.IAzureAccountService {
...
 	public async getAccounts(): Promise<mssql.IAccount[]> {
 		return await this._accountStore.getAccounts();
 	}
+
 	public async getAccountSecurityToken(account: mssql.IAccount, tenantId: string | undefined): Promise<mssql.Token> {
 		return await this._azureController.getAccountSecurityToken(account, tenantId, providerSettings.resources.azureManagementResource);
 	}
+
+	/**
+	 * Returns Azure subscriptions with tenant and token for each given account
```
Comment (Charles-Gagnon) [inline src/services/azureAccountService.ts:34]: Shouldn't this be `Returns Azure account sessions...`?

File: src/services/azureResourceService.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import providerSettings from '../azure/providerSettings';
+
+export class AzureResourceService implements mssql.IAzureResourceService {
+
+	private _accountStore: AccountStore;
+	constructor(
+		private _azureController: AzureController,
+		private _azureResourceController: AzureResourceController,
+		private _context: vscode.ExtensionContext) {
+		this._accountStore = new AccountStore(this._context);
```
Comment (Charles-Gagnon) [inline src/services/azureResourceService.ts:23]: We shouldn't be making our own account store. We should be using the one from the azure account service.

File: src/services/azureResourceService.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	 public async createOrUpdateServer(session: mssql.IAzureAccountSession, resourceGroupName: string, serverName: string, parameters: Server): Promise<string | undefined> {
+		await this.checkAndRefreshToken(session);
+		return await this._azureResourceController.createOrUpdateServer(session.subscription.subscriptionId, resourceGroupName, serverName, parameters, session.token);
+	}
+
+	/**
+	 * Verifies if the token still valid, refreshes the token for given account
+	 * @param session
+	 */
+	private async checkAndRefreshToken(session: mssql.IAzureAccountSession) {
```
Comment (Charles-Gagnon) [inline src/services/azureResourceService.ts:54]: This should be a helper method in the azure controller

File: src/azure/azureController.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply src/azure/azureController.ts:321]: It seems like this is for testing purposes. It's much better to use the DI pattern where you inject a factory method into the constructor - something like this : 

```
export type SubscriptionClientFactory = (token: string) => string;
function defaultSubscriptionClientFactory(token: string) {
	return new SubscriptionClient(new TokenCredentialWrapper(token));
}
...
constructor(private _subscriptionClientFactory: SubscriptionClientFactory = defaultSubscriptionClientFactory) {
```

And then whenever you need to get the client you just call `this._subscriptionClientFactory(token)`

This removes a lot of the "extra" code and is a lot easier to understand since you're using the factory in the same way regardless of how it was implemented.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/azure/azureController.ts (hunk 3 above)
Comment (llali) [reply src/azure/azureController.ts:255]: if you are asking about why the function is here, I thought it's related to account and that's why I put it here. I was debating about having this function in this API or resources API and I was not sure at the end. as you see, in extension APIs, I put this in account service not resource service. a=if you think it make more sense to put it in resources, I can move it

File: src/azure/azureResourceController.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+			const locationsPages = await subClient.subscriptions.listLocations(session.subscription.subscriptionId);
+			let locations = await azureUtils.getAllValues(locationsPages, (v) => v);
+			return locations.sort((a, b) => (a.name || '').localeCompare(b.name || ''));
+		} else {
+			throw new Error('Invalid session');
+		}
+	}
+
+	/**
+	 * Creates or updates a Azure SQL server for given subscription, resource group and location
```
Comment (Charles-Gagnon) [inline src/azure/azureResourceController.ts:35]: Please add doc comments for params and return statement

File: typings/vscode-mssql.d.ts [hunk 11]
```diff
@@ -437,6 +452,29 @@ declare module 'vscode-mssql' {
...
+		/**
+		 * Returns Azure subscriptions with tenant and token for each given account
+		 */
+		getAccountSessions(account: IAccount): Promise<IAzureAccountSession[]>;
+	}
+
+	export interface IAzureResourceService {
+
+		/**
+		 * Returns Azure resource groups for given subscription
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:465]: Please add doc comments for all these for the params and return values

**Review by Charles-Gagnon — APPROVED** [review]
Few minor things but otherwise looks good! Thanks for taking the time to split this out.

File: src/azure/azureResourceController.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	 * Returns Azure locations for given session
+	 */
+	public async getLocations(session: mssql.IAzureAccountSession): Promise<Location[]> {
+		const subClient = this._subscriptionClientFactory(session.token);
+		if (session.subscription?.subscriptionId) {
+			const locationsPages = await subClient.subscriptions.listLocations(session.subscription.subscriptionId);
+			let locations = await azureUtils.getAllValues(locationsPages, (v) => v);
+			return locations.sort((a, b) => (a.name || '').localeCompare(b.name || ''));
+		} else {
+			throw new Error('Invalid session');
```
Comment (kisantia) [inline src/azure/azureResourceController.ts:30]: localize

**Review by kisantia — APPROVED** [review]

File: src/azure/azureController.ts [hunk 13]
```diff
@@ -239,6 +251,30 @@ export class AzureController {
 		}
 	}
 
+	/**
+	 * Returns Azure subscriptions with tenant and token for each given account
```
Comment (alanrenmsft) [inline src/azure/azureController.ts:255]: comment doesn't match the method name, also in the method body, you used subscriptions but the type is session, Subscription == Session?

File: src/azure/azureResourceController.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+	/**
+	 * Creates or updates a Azure SQL server for given subscription, resource group and location
+	 */
+	public async createOrUpdateServer(
+		subscriptionId: string,
+		resourceGroupName: string,
+		serverName: string,
+		parameters: Server,
+		token: mssql.Token): Promise<string | undefined> {
```
Comment (abist) [inline src/azure/azureResourceController.ts:42]: this should be `Promise<string> | undefined`

File: src/azure/azureController.ts [hunk 15]
```diff
@@ -279,4 +315,20 @@ export class AzureController {
...
 
+
+	/**
+	 * Verifies if the token still valid, refreshes the token for given account
+	 * @param session
+	 */
+	public async checkAndRefreshToken(
+		session: mssql.IAzureAccountSession,
+		accountStore: AccountStore): Promise<void> {
+		const currentTime = new Date().getTime() / 1000;
```
Comment (alanrenmsft) [inline src/azure/azureController.ts:326]: what does / 1000 mean?

File: src/azure/azureResourceController.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+		if (subscriptionId && resourceGroupName) {
+			const sqlClient = this._sqlManagementClientFactory(token, subscriptionId);
+			if (sqlClient) {
+				const result = await sqlClient.servers.beginCreateOrUpdateAndWait(resourceGroupName,
+					serverName, parameters);
+
+				return result.fullyQualifiedDomainName;
```
Comment (alanrenmsft) [inline src/azure/azureResourceController.ts:49]: what if the result is undefined?

File: src/azure/azureResourceController.ts (hunk 14 above)
Comment (Charles-Gagnon) [reply src/azure/azureResourceController.ts:42]: That's not correct. It always returns a promise (because it's async). The wrapped type is either `string | undefined`. So what she has is correct.

File: src/azure/azureResourceController.ts (hunk 12 above)
Comment (llali) [reply src/azure/azureResourceController.ts:30]: I noticed that we didn't localize the error coming from extension API. I thought it might be intentional, so I did the same

File: src/azure/azureController.ts (hunk 15 above)
Comment (llali) [reply src/azure/azureController.ts:326]: copy the code somewhere else in mssql extension which we refresh token when it's expired. getTime is in milliseconds and to compare to expire it needs to get converted

File: src/azure/azureResourceController.ts (hunk 16 above)
Comment (llali) [reply src/azure/azureResourceController.ts:49]: I don't think that happens. Azure APIs throw error if fails to create server

**Review by alanrenmsft — APPROVED** [review]
